### ISSUE TITLE
Model Fails to Learn in Gymnasium Environment (e.g., LunarLander-v3)

### ISSUE THREAD

**Issue by Shengqi-Kong** [alex-petrenko/sample-factory#317]
I am new to the Sample Factory framework, so I followed the getting started documentation to experiment with Gymnasium environments like **`LunarLander-v3.`**

I used the example code that registers `LunarLander-v3`using `register_env`, and kept the default model settings. However, during training, I found that the learning performance was very poor. After around 30 minutes of training, there was almost no improvement—the average reward remained between `-200 and -190`.

First of all, I suspected that the issue might be related to the default model. So I looked into the source code and found that the default model is designed as follows:

- If the input has multiple channels, the model uses a convolutional network.
- Otherwise, it uses a fully connected network.

So it’s probably not a model-related problem in this case. 

Below is the training code I used. It’s almost the same as the one in sf_examples/train_gym_env.py:

```python
import sys
from typing import Optional

import gymnasium as gym

from sample_factory.cfg.arguments import parse_full_cfg, parse_sf_args
from sample_factory.envs.env_utils import register_env
from sample_factory.train import run_rl


def make_gym_env_func(full_env_name, cfg=None, env_config=None, render_mode: Optional[str] = None):
    return gym.make(full_env_name, render_mode=render_mode)


def register_custom_components():
    register_env("LunarLander-v3", make_gym_env_func)


def parse_custom_args(argv=None, evaluation=False):
    parser, cfg = parse_sf_args(argv=argv, evaluation=evaluation)
    cfg = parse_full_cfg(parser, argv)
    return cfg


def main():
    """Script entry point."""
    register_custom_components()
    custom_argv = ['--env=LunarLander-v3','--experiment=example_LunarLanderv2']
    cfg = parse_custom_args(argv=custom_argv)
    status = run_rl(cfg)
    return status


if __name__ == "__main__":
    sys.exit(main())

```
And the console infomation is like follows:
```
[2025-04-17 19:55:10,006][11424] Updated weights for policy 0, policy_version 6030 (0.0007)
[2025-04-17 19:55:11,032][11424] Updated weights for policy 0, policy_version 6040 (0.0008)
[2025-04-17 19:55:11,943][11424] Updated weights for policy 0, policy_version 6050 (0.0007)
[2025-04-17 19:55:12,867][11424] Updated weights for policy 0, policy_version 6060 (0.0008)
[2025-04-17 19:55:13,856][11424] Updated weights for policy 0, policy_version 6070 (0.0009)
[2025-04-17 19:55:14,754][11066] Fps is (10 sec: 10649.7, 60 sec: 10683.7, 300 sec: 10479.5). Total num frames: 6224896. Throughput: 0: 10637.5. Samples: 6217524. Policy #0 lag: (min: 0.0, avg: 1.4, max: 3.0)
[2025-04-17 19:55:14,755][11066] Avg episode reward: [(0, '-198.289')]
[2025-04-17 19:55:14,792][11424] Updated weights for policy 0, policy_version 6080 (0.0008)
[2025-04-17 19:55:15,808][11424] Updated weights for policy 0, policy_version 6090 (0.0008)
[2025-04-17 19:55:16,822][11424] Updated weights for policy 0, policy_version 6100 (0.0008)
[2025-04-17 19:55:17,723][11424] Updated weights for policy 0, policy_version 6110 (0.0007)
[2025-04-17 19:55:18,758][11424] Updated weights for policy 0, policy_version 6120 (0.0008)
[2025-04-17 19:55:19,677][11424] Updated weights for policy 0, policy_version 6130 (0.0008)
[2025-04-17 19:55:19,754][11066] Fps is (10 sec: 10547.5, 60 sec: 10649.6, 300 sec: 10476.0). Total num frames: 6277120. Throughput: 0: 10608.5. Samples: 6248538. Policy #0 lag: (min: 0.0, avg: 1.4, max: 3.0)
[2025-04-17 19:55:19,755][11066] Avg episode reward: [(0, '-202.334')]


```

I'd really appreciate any help or insights. Thanks in advance!

**Comment by Shengqi-Kong** [comment]
Hi, I noticed something strange during my experiments.

I used Sample Factory to successfully train my model on a native Ubuntu system (dual-boot setup). However, when I tried running the same training setup on WSL2 (Windows Subsystem for Linux), the training failed.

Both environments are configured similarly, so I'm not sure what's causing the issue on WSL2. Has anyone encountered similar problems or have suggestions for debugging?


```
Name: sample-factory
Version: 2.1.1
```


```
[2025-04-18 17:24:57,780][51667] Fps is (10 sec: 17100.8, 60 sec: 17152.0, 300 sec: 16805.3). Total num frames: 2941952. Throughput: 0: 17125.4. Samples: 2932507. Policy #0 lag: (min: 0.0, avg: 1.3, max: 3.0)
[2025-04-18 17:24:57,781][51667] Avg episode reward: [(0, '277.319')]
[2025-04-18 17:24:57,781][52019] Saving new best policy, reward=277.319!
[2025-04-18 17:24:58,165][52136] Updated weights for policy 0, policy_version 2880 (0.0006)
[2025-04-18 17:24:58,762][52136] Updated weights for policy 0, policy_version 2890 (0.0007)
[2025-04-18 17:24:59,359][52136] Updated weights for policy 0, policy_version 2900 (0.0007)
[2025-04-18 17:24:59,974][52136] Updated weights for policy 0, policy_version 2910 (0.0008)
[2025-04-18 17:25:00,579][52136] Updated weights for policy 0, policy_version 2920 (0.0007)
[2025-04-18 17:25:01,182][52136] Updated weights for policy 0, policy_version 2930 (0.0007)
[2025-04-18 17:25:01,784][52136] Updated weights for policy 0, policy_version 2940 (0.0007)
[2025-04-18 17:25:02,401][52136] Updated weights for policy 0, policy_version 2950 (0.0006)
[2025-04-18 17:25:02,781][51667] Fps is (10 sec: 17100.9, 60 sec: 17134.9, 300 sec: 16810.7). Total num frames: 3026944. Throughput: 0: 17120.8. Samples: 2983587. Policy #0 lag: (min: 0.0, avg: 1.4, max: 3.0)
[2025-04-18 17:25:02,781][51667] Avg episode reward: [(0, '263.860')]
[2025-04-18 17:25:02,998][52136] Updated weights for policy 0, policy_version 2960 (0.0008)
[2025-04-18 17:25:03,603][52136] Updated weights for policy 0, policy_version 2970 (0.0007)
[2025-04-18 17:25:04,199][52136] Updated weights for policy 0, policy_version 2980 (0.0008)
[2025-04-18 17:25:04,803][52136] Updated weights for policy 0, policy_version 2990 (0.0007)
[2025-04-18 17:25:05,402][52136] Updated weights for policy 0, policy_version 3000 (0.0006)
[2025-04-18 17:25:05,985][52136] Updated weights for policy 0, policy_version 3010 (0.0007)
[2025-04-18 17:25:06,595][52136] Updated weights for policy 0, policy_version 3020 (0.0006)
[2025-04-18 17:25:07,198][52136] Updated weights for policy 0, policy_version 3030 (0.0008)
[2025-04-18 17:25:07,781][51667] Fps is (10 sec: 16998.4, 60 sec: 17100.8, 300 sec: 16815.7). Total num frames: 3111936. Throughput: 0: 17112.9. Samples: 3085657. Policy #0 lag: (min: 0.0, avg: 1.3, max: 3.0)
[2025-04-18 17:25:07,781][51667] Avg episode reward: [(0, '271.080')]
[2025-04-18 17:25:07,814][52136] Updated weights for policy 0, policy_version 3040 (0.0007)
```

The above result was obtained in less than 5 minutes on my native Linux (dual-boot) system.

I'm wondering if Sample Factory depends on any features that are only available in a native Linux environment, and therefore might not work properly on WSL2?

**Comment by alex-petrenko** [comment]
Hey @Shengqi-Kong 
When you say the training has failed, what did you mean by that? Does it fail with an error or does it get low rewards?

How confident are you that this is not random seed issue, e.g. if you run training on Linux 5 times, does it consistently get these higher rewards?

I do not have much experience with WSL2, so you might be on your own here. Here's a quick question though, if you run an untrained policy without any learning, do they get similar rewards? I think it's conceivable that there some WSL-specific issue with shared memory or something that causes weights not to update, although it does not seem likely (you'd probably see some sort of crash instead)

**Comment by Shengqi-Kong** [comment]
Thank you for your response! I realized I had overlooked a few details:

1. By `"Failed Training"`,  I meant that the agent was consistently receiving low rewards, indicating ineffective learning.
2. Regarding the random seed, I haven't explored that aspect yet.
3. Initially, I trained the model on `WSL2`. Since the Gymnasium environment is relatively simple, I expected the training to be fast. However, even after `30 minutes` of training, the average reward remained quite low. The next day, I switched to a `native Ubuntu system` — and it worked! Within just 5 minutes, the reward began to increase rapidly, and the trained model started performing effectively.

Thanks again for your help. For anyone else experiencing issues with training on WSL2, I recommend trying it on a native Linux system.

**Comment by JulioSanchezD** [comment]
I can confirm what @Shengqi-Kong is saying, it happens the same for me with vizdoom, health-gathering env, agent cannot learn at all, totally random moves even after 1B steps. I replicated the same code and environment on a google colab and works just fine there.
